### PR TITLE
[Tests-Only] Refactor webdav auth

### PR DESCRIPTION
**Pull request by kiranparajuli589**

### CODE REVIEW

File: tests/acceptance/features/apiAuthWebDav/webDavMKCOLAuth.feature [hunk 1]
```diff
@@ -12,85 +12,93 @@ Feature: create folder using MKCOL
...
-      | /remote.php/dav/files/%username%/PARENT            | 403       |      |
-      | /remote.php/dav/files/%username%/PARENT/parent.txt | 409       |      |
-      | /remote.php/dav/files/%username%/does-not-exist    | 403       |      |
+    When user "Brian" requests these endpoints with "MKCOL" including body "" about user "Alice"
+      | endpoint                                           |
+      | /remote.php/dav/files/%username%/textfile0.txt     |
+      | /remote.php/dav/files/%username%/PARENT            |
+      | /remote.php/dav/files/%username%/PARENT/parent.txt |
+      | /remote.php/dav/files/%username%/does-not-exist    |
+    Then the HTTP status code of responses on all endpoints should be "403"
```
Comment (phil-davis) [inline tests/acceptance/features/apiAuthWebDav/webDavMKCOLAuth.feature:44]: What about the 409 in the old table?

File: tests/acceptance/features/apiAuthWebDav/webDavPOSTAuth.feature [hunk 2]
```diff
@@ -13,85 +13,92 @@ Feature: get file info using POST
...
-      | /remote.php/webdav/textfile0.txt                   | 501       | doesnotmatter |
-      | /remote.php/dav/files/%username%/textfile1.txt     | 501       | doesnotmatter |
-      | /remote.php/dav/files/%username%/PARENT/parent.txt | 501       | doesnotmatter |
-      | /remote.php/webdav/PARENT                          | 501       | doesnotmatter |
-      | /remote.php/dav/files/%username%/FOLDER            | 501       | doesnotmatter |
+      | /remote.php/webdav/textfile0.txt                   |
+      | /remote.php/dav/files/%username%/textfile1.txt     |
+      | /remote.php/dav/files/%username%/PARENT/parent.txt |
+      | /remote.php/webdav/PARENT                          |
+      | /remote.php/dav/files/%username%/FOLDER            |
```
Comment (phil-davis) [inline tests/acceptance/features/apiAuthWebDav/webDavPOSTAuth.feature:104]: Needs `Then` step like:
```
    Then the HTTP status code of responses on all endpoints should be "501"
```

File: tests/acceptance/features/apiAuthWebDav/webDavPUTAuth.feature [hunk 3]
```diff
@@ -13,86 +13,94 @@ Feature: get file info using PUT
...
+      | endpoint                                           |
+      | /remote.php/webdav/textfile0.txt                   |
+      | /remote.php/dav/files/%username%/textfile1.txt     |
+      | /remote.php/dav/files/%username%/PARENT/parent.txt |
+      | /remote.php/webdav/PARENS                          |
+      | /remote.php/dav/files/%username%/FOLDERS           |
       # this folder already exists so gives 409 - CONFLICT
-      | /remote.php/dav/files/%username%/FOLDER            | 409       | doesnotmatter |
+      | /remote.php/dav/files/%username%/FOLDER            |
+    Then the HTTP status code of responses on all endpoints should be "204,201,409"
```
Comment (phil-davis) [inline tests/acceptance/features/apiAuthWebDav/webDavPUTAuth.feature:106]: I don't see code to handle a list of code here. And how will it know which of the 6 API  calls are supposed to have status 204, 201 and 409?

File: tests/acceptance/features/bootstrap/AuthContext.php [hunk 4]
```diff
@@ -239,17 +240,79 @@ public function userRequestsEndpointsWithBasicAuth($user, $method, TableNode $ta
...
+	public function userRequestsEndpointsWithBasicAuthAndGeneratedPasswordWithProperty(
+		$user, $method, $property, $ofUser, TableNode $table
+	) {
+		$this->requestEndpointsWithBasicAuthAndGeneratedPassword(
+			$user, $method, $ofUser, $table, null, $property
+		);
+	}
+
+	/**
+	 * @When the user :user requests these endpoints with :method with body :body using the basic auth and generated app password about user :ofUser
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/AuthContext.php:267]:
```suggestion
	 * @When the user :user requests these endpoints with :method with body :body using basic auth and generated app password about user :ofUser
```

File: tests/acceptance/features/bootstrap/AuthContext.php [hunk 5]
```diff
@@ -239,17 +240,79 @@ public function userRequestsEndpointsWithBasicAuth($user, $method, TableNode $ta
 	 * @throws Exception
 	 */
 	public function userRequestsEndpointsWithBasicAuthAndGeneratedPassword($user, $method, $ofUser, TableNode $table) {
+		$this->requestEndpointsWithBasicAuthAndGeneratedPassword($user, $method, $ofUser, $table);
+	}
+
+	/**
+	 * @When the user :user requests these endpoints with :method to (get|set) property :property using the basic auth and generated app password about user :ofUser
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/AuthContext.php:247]:
```suggestion
	 * @When the user :user requests these endpoints with :method to (get|set) property :property using basic auth and generated app password about user :ofUser
```

File: tests/acceptance/features/bootstrap/OCSContext.php [hunk 6]
```diff
@@ -530,30 +578,30 @@ public function userSendsRequestToTheseEndpointsWithBody($user, $method, $ofUser
 				$row['endpoint'],
 				$method,
 				$this->featureContext->getPasswordForUser($user),
-				$row['body']
+				$body
+			);
+			$this->featureContext->pushToLastStatusCodesArray(
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/OCSContext.php:583]: This method used to have an optional ocs-code. Do we still need to save that?

File: tests/acceptance/features/bootstrap/OCSContext.php [hunk 7]
```diff
@@ -563,13 +611,11 @@ public function userRequestsTheseEndpointsWithUsingThePasswordOfUser($asUser, $m
...
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/OCSContext.php:616]: This method used to have an optional ocs-code. Do we still need to save that?

**Review by phil-davis — COMMENTED** [review]
I noticed that there are also scenarios like these in apiAuthOCS suite:
ocsDELETEAuth
ocsGETAuth
ocsPUTAuth

Those look like they need similar refactoring.

File: tests/acceptance/features/apiAuthWebDav/webDavPUTAuth.feature [hunk 8]
```diff
@@ -42,6 +42,10 @@ Feature: get file info using PUT
       | /remote.php/dav/files/%username%/PARENTS            |
       | /remote.php/dav/files/%username%/PARENTS/parent.txt |
     Then the HTTP status code of responses on all endpoints should be "403"
+    When user "Brian" requests these endpoints with "PUT" including body "doesnotmatter" about user "Alice"
+      | endpoint                                            |
+      | /remote.php/dav/files/%username%/PARENTS/parent.txt |
```
Comment (phil-davis) [inline tests/acceptance/features/apiAuthWebDav/webDavPUTAuth.feature:47]: This looks the same as line 43 - I guess one of them will fail.

**Review by phil-davis — COMMENTED** [review]
Let's see what else CI tells us.
And then please rebase and squash whne making the last small fixes.

**Comment by kiranparajuli589** [comment]
@phil-davis addressed your comments, please review!

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/core/25942/50/13
```
  Scenario: using OCS with admin basic auth                                 # /drone/src/tests/acceptance/features/apiAuthOcs/ocsGETAuth.feature:235
    When the administrator requests these endpoint with "GET"               # AuthContext::adminRequestsEndpoint()
      | endpoint                 |
      | /ocs/v1.php/cloud/apps   |
      | /ocs/v1.php/cloud/groups |
      | /ocs/v1.php/cloud/users  |
    Then the HTTP status code of responses on all endpoints should be "200" # FeatureContext::theHTTPStatusCodeOfResponsesOnAllEndpointsShouldBe()
      │ array(1) {
      │   [0]=>
      │   int(200)
      │ }
      │ 
    And the OCS status code of responses on all endpoints should be "100"   # FeatureContext::theOCSStatusCodeOfResponsesOnAllEndpointsShouldBe()
    When the administrator requests these endpoint with "GET"               # AuthContext::adminRequestsEndpoint()
      | endpoint                 |
      | /ocs/v2.php/cloud/apps   |
      | /ocs/v2.php/cloud/groups |
      | /ocs/v2.php/cloud/users  |
    Then the HTTP status code of responses on all endpoints should be "200" # FeatureContext::theHTTPStatusCodeOfResponsesOnAllEndpointsShouldBe()
      │ array(1) {
      │   [0]=>
      │   int(200)
      │ }
      │ 
    And the OCS status code of responses on all endpoints should be "200"   # FeatureContext::theOCSStatusCodeOfResponsesOnAllEndpointsShouldBe()
      Expected same but found different ocs status codes of last requested responses (Exception)
```
It is going to be annoying not to know what were the different status codes returned (for OCS and similar for HTTP).

It would be nice to report the list of status codes. Because they are numbers, it will be easy and clear enough to put a comma-separated list of all the actual returned codes into the exception message. So it could say:

`Expected all responses to have OCS status 200 but got 200,201,404,200,200`

Then someone reading it can have a chance to "guess" which request(s) had the unexpected status.

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 9]
```diff
@@ -1787,7 +1787,10 @@ public function theHTTPStatusCodeOfResponsesOnAllEndpointsShouldBe($statusCode)
 				'Responses did not return expected http status code'
 			);
 		} else {
-			throw new \Exception('Expected same but found different http status codes of last requested responses');
+			throw new \Exception(
+				'Expected same but found different http status codes of last requested responses' .
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebDav.php:1791]:
```suggestion
				'Expected same but found different http status codes of last requested responses. ' .
```

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 10]
```diff
@@ -1808,7 +1811,10 @@ public function theOCSStatusCodeOfResponsesOnAllEndpointsShouldBe($statusCode) {
 				'Responses did not return expected ocs status code'
 			);
 		} else {
-			throw new \Exception('Expected same but found different ocs status codes of last requested responses');
+			throw new \Exception(
+				'Expected same but found different ocs status codes of last requested responses' .
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebDav.php:1815]:
```suggestion
				'Expected same but found different ocs status codes of last requested responses. ' .
```

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/core/25954/34/6
drone had some "random" yarn install errors.
I will let the acceptance tests finish so we can see any fails there, then restart drone.

File: tests/acceptance/features/bootstrap/AuthContext.php [hunk 11]
```diff
@@ -392,19 +467,20 @@ public function userRequestsTheseEndpointsUsingNewBrowserSession($method, TableN
 	 */
 	public function userRequestsEndpointsUsingTheGeneratedAppPasswordThenStatusCodeAboutUser($method, $user, TableNode $table) {
 		$user = \strtolower($this->featureContext->getActualUsername($user));
-		$this->featureContext->verifyTableNodeColumns($table, ['endpoint', 'http-code'], ['ocs-code']);
+		$this->featureContext->verifyTableNodeColumns($table, ['endpoint']);
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/AuthContext.php:470]: For future safety this step could do:
```
$this->featureContext->emptyLastHTTPStatusCodesArray();
```

Comment (kiranparajuli589) [reply tests/acceptance/features/bootstrap/AuthContext.php:470]: done!

**Comment by phil-davis** [comment]
@kiranparajuli589 https://drone.owncloud.com/owncloud/core/25956/51/13
```
  Scenario: send MKCOL requests to another user's webDav endpoints as normal user                # /drone/src/tests/acceptance/features/apiAuthWebDav/webDavMKCOLAuth.feature:37
    When user "Brian" requests these endpoints with "MKCOL" including body "" about user "Alice" # OCSContext::userSendsRequestToTheseEndpointsWithBody()
      | endpoint                                           |
      | /remote.php/dav/files/%username%/textfile0.txt     |
      | /remote.php/dav/files/%username%/PARENT            |
      | /remote.php/dav/files/%username%/PARENT/parent.txt |
    Then the HTTP status code of responses on all endpoints should be "403"                      # FeatureContext::theHTTPStatusCodeOfResponsesOnAllEndpointsShouldBe()
      Expected same but found different http status codes of last requested responses.Found status codes: 403,403,409 (Exception)
```
needs looking at.

After that, IMO the general code here is good, so squash all the commits and maybe this can be merged.

**Review by phil-davis — APPROVED** [review]

**Pull request merged**